python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,471 @@
+                try:
+                    tester_dag.add_edge(prereq, tester.getTestName())
+
+                # Handle Cyclic errors (tester status has already be applied)
+                except dag.DAGValidationError:
+                    pass
+
+            # Discover skipped tests
+            if not tester.getRunnable(self.options):
+                skipped_or_failed_tests.add(tester)
+
+        # Discover skipped tests and remove downstream tests that require this skipped test
+        for tester in skipped_or_failed_tests:
+            # Delete and return downstream nodes (prints 'skipped dependency' tests to the screen)
+            downstream_deleted = self.deleteDownstreamTests(tester_dag, tester)
+
+            # Delete this test
+            tester_dag.delete_node_if_exists(tester.getTestName())
+
+            # Inform the user it was skipped by adding it to the queue
+            self.statusGo(tester.getTestName())
+
+        # Get top level jobs we can launch
+        ind_jobs = tester_dag.ind_nodes()
+
+        # Maybe all jobs in this group got skipped?
+        if ind_jobs == []:
+            return
+
+        # Add our local DAG to the global shared dag
+        #with self.thread_lock:
+        self.shared_testers_dag[testers[-1].getTestDir()] = tester_dag
+
+        # Set a pending status on these jobs _before_ launching them
+        for job in ind_jobs:
+            self.scheduled_name_to_object[job].setStatus('QUEUED', self.scheduled_name_to_object[job].bucket_pending)
+
+        # Launch the jobs
+        for job in ind_jobs:
+            self.runnerGo(job)
+
+    # Launch jobs stored in our queues. Method is blocked until both queue(s) and active jobs are empty
+    def waitFinish(self):
+        while len(self.active_jobs) != 0 or not self.runner_queue.empty() or not self.status_queue.empty():
+            # Using a thread lock, check and do things every iteration
+            with self.thread_lock:
+                self.handleActiveTests()
+
+            # sleep for just a tick or two
+            sleep(0.1)
+
+        ### Runner Pool
+        # Wait for the runner pool to empty before waiting on the status pool
+        # The runner pool and status pool can exhange jobs between each other but it is always the status pool
+        # that will have items in the queue last (that last job with a finished status to be printed)
+
+        # Close the runner_pool so jobs can no longer be scheduled for launching
+        self.runner_pool.close()
+
+        # Wait for the runner_pool to empty
+        self.runner_pool.join()
+
+        # Close the status_pool so jobs can no longer be scheduled for status printing
+        self.status_pool.close()
+
+        # Wait for the status_pool to empty
+        self.status_pool.join()
+
+    # Method to do things needed to actively running tests.
+    # NOTE: We are in a thread locked state, so be quick
```

Comment:
Thanks for the comment here, but I'd avoid locking around a whole routine. It's actually two or three routines down before you actually do anything with the data structure. You should only need to lock the actual changes to the structure.